File: Examples/TS/test7.ts
type MyList<A> = MyNil<A> | MyCons<A>;
interface MyNil<A> {
}
interface MyCons<A> {
    hd: A;
    tl: MyList<A>;
}
type Colour = Red | Black | White;
interface Red{
}
interface Black{
}
interface White{
}
type Tree<A, B> = Leaf<A, B> | Node2<A, B>;
interface Leaf<A, B> {
    value: A;
}
interface Node2<A, B> {
    value: B;
    left: Tree<A, B>;
    right: Tree<A, B>;
}
interface Point{
    x: number;
    y: number;
}
interface Pair<A, B> {
    fst: A;
    snd: B;
}
const p: Pair<number, string> = { kind: "Pair", value: { fst: 4, snd: "something" } }
const p1 = p.fst
const p2 = p.snd
function a(): number {
    if (p.kind === "Pair") {
        const fst: f, snd: s = p.value;
        
        return f;
    }
}
function b(): string {
    if (p.kind === "Pair") {
        const fst: f, snd: s = p.value;
        
        return s;
    }
}
function toInt(c: Colour): number {
    if (c.kind === "Red") {
        const  = c.value;
        
        return 0;
    } else if (c.kind === "Black") {
        const  = c.value;
        
        return 1;
    } else if (c.kind === "White") {
        const  = c.value;
        
        return 2;
    }
}
function length<A>(l: MyList<A>): number {
    if (l.kind === "MyNil") {
        const  = l.value;
        
        return 0;
    } else if (l.kind === "MyCons") {
        const hd: h, tl: t = l.value;
        
        return (1 + length(t));
    }
}
function max(x: number, y: number): number {
    if (x > y) {
        return x;
    } else {
        return y;
    }
}
function size<A, B>(t: Tree<A, B>): number {
    if (t.kind === "Leaf") {
        const value: _ = t.value;
        
        return 1;
    } else if (t.kind === "Node2") {
        const value: _, left: l, right: r = t.value;
        
        return max(size(l), size(r));
    }
}
function copy(t: Tree<number, string>): Tree<number, string> {
    if (t.kind === "Leaf") {
        const value: a = t.value;
        
        return { kind: "Leaf", value: { value: a } };
    } else if (t.kind === "Node2") {
        const value: b, left: l, right: r = t.value;
        
        const l2 = copy(l)
        const r2 = copy(r)
        return { kind: "Node2", value: { value: b, left: l2, right: r2 } };
    }
}
const c: Colour = { kind: "Red", value: {  } }
const l: Tree<number, string> = { kind: "Leaf", value: { value: 3 } }
const t: Tree<number, string> = { kind: "Node2", value: { value: "abc", left: { kind: "Leaf", value: { value: 7 } }, right: { kind: "Leaf", value: { value: size(l) } } } }

File: Examples/PY/test7.py
from dataclasses import dataclass
type MyList[A] = MyNil[A] | MyCons[A]
@dataclass
class MyNil[A]:
    pass
@dataclass
class MyCons[A]:
    hd: A
    tl: MyList[A]
type Colour = Red | Black | White
@dataclass
class Red:
    pass
@dataclass
class Black:
    pass
@dataclass
class White:
    pass
type Tree[A, B] = Leaf[A, B] | Node2[A, B]
@dataclass
class Leaf[A, B]:
    value: A
@dataclass
class Node2[A, B]:
    value: B
    left: Tree[A, B]
    right: Tree[A, B]
@dataclass
class Point:
    x: int
    y: int
@dataclass
class Pair[A, B]:
    fst: A
    snd: B
p: Pair[int, str] = Pair(fst=4, snd="something")
p1 = p.fst
p2 = p.snd
def a() -> int:
    match p:
        case Pair(fst=f, snd=s):
            return f
def b() -> str:
    match p:
        case Pair(fst=f, snd=s):
            return s
def toInt(c: Colour) -> int:
    match c:
        case Red():
            return 0
        case Black():
            return 1
        case White():
            return 2
def length[A](l: MyList[A]) -> int:
    match l:
        case MyNil():
            return 0
        case MyCons(hd=h, tl=t):
            return 1 + (length(t))
def max(x: int, y: int) -> int:
    if x > y:
        return x
    else:
        return y
def size[A, B](t: Tree[A, B]) -> int:
    match t:
        case Leaf(value=_):
            return 1
        case Node2(value=_, left=l, right=r):
            return max(size(l), size(r))
def copy(t: Tree[int, str]) -> Tree[int, str]:
    match t:
        case Leaf(value=a):
            return Leaf(value=a)
        case Node2(value=b, left=l, right=r):
            l2 = copy(l)
            r2 = copy(r)
            return Node2(value=b, left=l2, right=r2)
c: Colour = Red()
l: Tree[int, str] = Leaf(value=3)
t: Tree[int, str] = Node2(value="abc", left=Leaf(value=7), right=Leaf(value=size(l)))
